File: src/main/java/io/github/mh321productions/serverapi/module/realtime/RealtimeModule.java
package io.github.mh321productions.serverapi.module.realtime;

import java.util.ArrayList;
import java.util.Collection;

import io.github.mh321productions.serverapi.Main;
import io.github.mh321productions.serverapi.api.APIImplementation;
import io.github.mh321productions.serverapi.api.SubPlugin;
import org.bukkit.World;

import io.github.mh321productions.serverapi.module.Module;
import io.github.mh321productions.serverapi.module.ModuleStopFunction;
import io.github.mh321productions.serverapi.module.ModuleType;

public class RealtimeModule extends Module {
	
	private RealtimeWorker worker = null;
	ArrayList<World> worlds;

	public RealtimeModule(Main plugin, APIImplementation api) {
		super(ModuleType.Realtime, plugin, api);
	}

	@Override
	protected boolean init() {
		worlds = new ArrayList<>();
		worker = new RealtimeWorker(this);
		worker.runTaskTimer(plugin, 0, 20);
		
		return true;
	}

	@Override
	protected void stopIntern() {
		if (worker != null) worker.cancel();
	}

	@Override
	public boolean registerSubPlugin(SubPlugin sub, ModuleStopFunction func) {
		return addIntern(sub, func);
	}

	@Override
	public void unregisterSubPlugin(SubPlugin sub) {
		removeIntern(sub);
	}
	
	public void addWorld(World w) {
		if (w != null) worlds.add(w);
	}
	
	public void addWorld(Collection<World> w) {
		if (w != null) worlds.addAll(w);
	}
	
	public void removeWorld(World w) {
		worlds.remove(w);
	}
	
	public void removeWorld(Collection<World> w) {
		if (w != null) worlds.removeAll(w);
	}

}

File: src/main/kotlin/io/github/mh321productions/serverapi/module/realtime/RealtimeModule.kt
package io.github.mh321productions.serverapi.module.realtime

import io.github.mh321productions.serverapi.Main
import io.github.mh321productions.serverapi.api.APIImplementation
import io.github.mh321productions.serverapi.api.SubPlugin
import io.github.mh321productions.serverapi.module.Module
import io.github.mh321productions.serverapi.module.ModuleType
import org.bukkit.World

class RealtimeModule(plugin: Main, api: APIImplementation) : Module(ModuleType.Realtime, plugin, api) {

    private lateinit var worker: RealtimeWorker
    var worlds = mutableListOf<World>()

    override fun init(): Boolean {
        worker = RealtimeWorker(this)
        worker.runTaskTimer(plugin, 0, 20)

        return true
    }

    override fun stopIntern() {
        worker.cancel()
    }

    override fun registerSubPlugin(sub: SubPlugin, func: () -> Unit): Boolean {
        return addIntern(sub, func)
    }

    override fun unregisterSubPlugin(sub: SubPlugin) {
        removeIntern(sub)
    }

    fun addWorld(w: World) {
        worlds.add(w)
    }

    fun addWorld(w: Collection<World>) {
        worlds.addAll(w)
    }

    fun removeWorld(w: World) {
        worlds.remove(w)
    }

    fun removeWorld(w: Collection<World>) {
        worlds.removeAll(w)
    }
}
